velox/functions/prestosql/window/AggregateWindow.cpp

```diff
@@ -0,0 +1,199 @@
+      // input rows from frameStart to frameEnd in the SelectivityVector.
+      // TODO : Try to re-use previous computations by advancing and retracting
+      // the aggregation based on the frame changes with each row. This would
+      // require adding new APIs to the Aggregate framework.
+      aggregate_->clear();
+      aggregate_->initializeNewGroups(&singleGroupRow_, singleGroup);
+
+      auto frameStartIndex = frameStartsVector[i] - minFrame_;
+      auto frameEndIndex = frameEndsVector[i] - minFrame_ + 1;
+      rows_.setValidRange(frameStartIndex, frameEndIndex, true);
+      rows_.updateBounds();
+
+      aggregate_->addSingleGroupRawInput(
+          singleGroupRow_, rows_, argVectors_, false);
+      aggregate_->extractValues(&singleGroupRow_, 1, &aggregateResultVector_);
+
+      result->copy(aggregateResultVector_.get(), resultOffset + i, 0, 1);
+
+      rows_.setValidRange(frameStartIndex, frameEndIndex, false);
+      rows_.updateBounds();
+    }
+  }
+
+ private:
+  void fillArgVectors(
+      const vector_size_t* frameStarts,
+      const vector_size_t* frameEnds,
+      vector_size_t numRows) {
+    // Argument values are needed to compute the aggregate. The aggregate
+    // computation for each row needs the values from its frameStart to frameEnd
+    // indices. So for each apply() we read arguments from the least
+    // frameStart to the highest frameEnd value in the row block.
+    minFrame_ = frameStarts[0];
+    maxFrame_ = frameEnds[0];
+    for (int i = 1; i < numRows; i++) {
+      minFrame_ = std::min(minFrame_, frameStarts[i]);
+      maxFrame_ = std::max(maxFrame_, frameEnds[i]);
+    }
+
+    vector_size_t numFrameRows = maxFrame_ + 1 - minFrame_;
+    for (int i = 0; i < argIndices_.size(); i++) {
+      argVectors_[i]->resize(numFrameRows);
+      partition_->extractColumn(
+          argIndices_[i], minFrame_, numFrameRows, 0, argVectors_[i]);
+    }
+  }
+
+  // Aggregate function object required for this window function evaluation.
+  std::unique_ptr<exec::Aggregate> aggregate_;
+
+  // Current WindowPartition used for accessing rows in the apply method.
+  const exec::WindowPartition* partition_;
+
+  // Args information : their types, column indexes in inputs and vectors
+  // used to populate values to pass to the aggregate function.
+  std::vector<TypePtr> argTypes_;
+  std::vector<column_index_t> argIndices_;
+  std::vector<VectorPtr> argVectors_;
+
+  // When computing for apply() function we work on one block of output at a
+  // time. For processing those rows we need arguments from the min frameStart
+  // index to the max frameEnd index values in that block.
+  vector_size_t minFrame_, maxFrame_;
+
+  // This is a single aggregate row needed by the aggregate function for its
+  // computation.
+  char* singleGroupRow_;
+  vector_size_t singleGroupRowSize_;
+
+  // Used for per-row aggregate computations.
+  SelectivityVector rows_;
```

Comment:
No need to reuse SelectivityVector across apply calls. The call itself is so expensive, that savings from reuse of SelectivityVector won't be visible. At the same time, not reusing SelectivityVector makes the code a bit simpler.